pkg/promclient/multi_api.go

```diff
@@ -101,6 +103,34 @@ type MultiAPI struct {
 	requiredCount   int // number "per key" that we require to respond
 }
 
+// Starts a memory monitoring goroutine that oversees the memory usage of the query and cancels memory intensive ones
+// Takes in the context cancel function and outputs the stop channel that needs to closed gracefully
+func (m *MultiAPI) monitorMemory(cancel func()) chan struct{} {
+	stopchan := make(chan struct{})
+	go func() {
+		d := 5 * time.Second
+		maxMem := uint64(5000)
+		for x := range time.Tick(d) {
+			select {
+			default:
+				var mem runtime.MemStats
+				runtime.ReadMemStats(&mem)
+				memAlloc := mem.Sys / 1024 / 1024
+				fmt.Printf("Sys = %v MiB  \t %v\n", (mem.Sys / 1024 / 1024), x)
```

Comment:
Print statements for debugging purposes, will be removed.
Idea: Print memory usage by each query as part of the debug flag. Should be simple to integrate if there's a use case for it.